### PR TITLE
Make train loss calculation optional

### PR DESCRIPTION
**Pull request by franchuterivera**
**What?**
Don't calculate the training loss unless required -- this helps with memory consumption in Auto-sklearn

 **Details**
 I found no strong reason to keep Y_actual_train -- maybe this is legacy.
Y_train_targets/Y_train_pred[*] variable (don't confuse with self.Y_train_targets) was unnecessarily created. Maybe it makes sense to enable a more strict mypy :).

### CODE REVIEW

**Review by eddiebergman — COMMENTED** [review]
While I agree this increases speed, I have three commets:
* By default I think we should have train_loss computed for explainability purposes
* Using `disable_evaluator_output` for this functionality means not introducing a new parameter to an already large list
* Requires a test

### Explainability over Optimization
Namely, I would rather explainability over optimizing for speed with `compute_train_loss = True` as the default. People who are new to autosklearn might use some of the explainability features or try poke at autosklearn to see what's going on, i.e. `leaderboard()` or other planned accessibility changes.
It also keeps auto-sklearn's behavior consistent with how it is already, leaving the speed up as a new opt-in feature.

### New Parameter vs. Reuse 
Changing this to default `True` means we can actually remove this explicit parameter and have it be a possible entry to `disable_evaluator_output`, i.e. `disable_evaluator_output = ['train_loss']`. 
I think this is more clear from a user perspective that they are disabling output being generated. I could imagine a context where a user heard 'to disable the training loss being calculated' and then wonders why there's a specific param for that but then also 'disable_evaluator_output` without knowing the specifics of how it works. Perhaps this requires renaming `disable_evaluator_output` to just `disable_output` though.

### Test
An parameterized standalone test would also need to be included. I think it's worth increasing the total testing time for it.

**Comment by mfeurer** [comment]
> Namely, I would rather explainability over optimizing for speed with compute_train_loss = True as the default.

I agree that explainability is usually a good thing, but how often is it important to look at the training loss? I agree that the user will want to look at the validation loss (and potentially also the test loss), but so far we don't even document that we compute the training loss.

> Perhaps this requires renaming disable_evaluator_output to just disable_output though.

Or `disable_internal_output`.

**Comment by eddiebergman** [comment]
> > Namely, I would rather explainability over optimizing for speed with compute_train_loss = True as the default.
> 
> I agree that explainability is usually a good thing, but how often is it important to look at the training loss? I agree that the user will want to look at the validation loss (and potentially also the test loss), but so far we don't even document that we compute the training loss.
> 
> > Perhaps this requires renaming disable_evaluator_output to just disable_output though.
> 
> Or `disable_internal_output`.

Good point that we don't document that at all. If it's to be disabled by default and we don't want to introduce another argument then `disable_internal_output` would be required to include it by default, i.e. `disable_internal_output = ['train_loss']`

**Comment by eddiebergman** [comment]
Ello @franchuterivera,

We decided that disabling train predictions by default is a good way to. For larger datasets and especially with a high number of CV folds, this could consume a lot of unnecessary memory.

We had some comments on perhaps some memory being consumed even with this patch but upon re-reviewing it, it seems all good to me. The main change I think would make sense is to change `disable_evaluator_output = False` to be `disable_evaluator_output = ['training_predictions']`. This way, there is no need to introduce a new parameter.

One more note that's not actionable but just a comment, this is a breaking change to the behaviour of auto-sklearn and anyone relying on training-losses for some purpose will have to be aware of these changes once updating so we should be careful to highlight it.

**Comment by mfeurer** [comment]
I just realized that we document the training loss in https://automl.github.io/auto-sklearn/master/examples/40_advanced/example_get_pipeline_components.html as requested by #670, so we should probably update that piece of documentation then.

**Comment by eddiebergman** [comment]
While reviewing this code, we could save further memory requirements. Not required for this PR but leaving this as a note.

Currently, the `TrainEvaluator` has a simplified loop:
```python
def fit_predict_and_loss(...):
   Y_train_pred = []
   Y_optimization_pred = []
   Y_valid_pred = []
   Y_test_pred = []
   train_splits = []
   for i in num_cv_folds:
      # ... populate each list
      
   # process all of them
```

We could save memory by doing each one individually
```python
   Y_test_preds = []
   for i in num_cv_folds:
       # populate
   # process
   del Y_test_preds
   
   # ...next ones
```

Or more succinctly,
```python
pred_types = [(info_for_y_test_preds), ..., (info_for_y_opt_preds)]
if self.compute_train_losses:
    pred_types += [(info_for_y_train_preds)]
    
for ... in pred_types:
   preds = []
   for i in num_cv_folds:
       # populate
   # process
```
